cpp/dolfinx/io/XDMFFile.cpp

```diff
@@ -47,72 +53,161 @@ std::string get_hdf5_filename(std::string filename)
 } // namespace
 
 //-----------------------------------------------------------------------------
-XDMFFile::XDMFFile(MPI_Comm comm, const std::string filename, Encoding encoding)
-    : _mpi_comm(comm), _filename(filename), _xml_doc(new pugi::xml_document),
-      _encoding(encoding)
+XDMFFile::XDMFFile(MPI_Comm comm, const std::string filename,
+                   const std::string file_mode, const Encoding encoding)
+    : _mpi_comm(comm), _filename(filename), _file_mode(file_mode),
+      _xml_doc(new pugi::xml_document), _encoding(encoding)
 {
-  // Open files here?
+  //
+  // Handle HDF5 and XDMF files with the file mode
+  // At the end of this we will have _hdf5_file and _xml_doc
+  // both pointing to a valid and opened file handles
+  //
+
+  if (_encoding == Encoding::HDF5)
+  {
+    // See https://www.hdfgroup.org/hdf5-quest.html#gzero on zero for
+    // _hdf5_file_id(0)
+
+    const std::string hdf5_filename = get_hdf5_filename(_filename);
+
+    // Open HDF5 file
+    const bool mpi_io = MPI::size(_mpi_comm.comm()) > 1 ? true : false;
+#ifndef H5_HAVE_PARALLEL
```

Comment:
No need for this check - we insist on parallel HDF5 in the top level CMake file.